contracts/Router.sol

```diff
@@ -0,0 +1,188 @@
+pragma solidity ^0.6.6;
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+
+
+interface IERC20 {
+    function balanceOf(address account) external view returns (uint256);
+    function transfer(address recipient, uint256 amount) external returns (bool);
+    function transferFrom(address sender, address recipient, uint256 amount) external returns (bool);
+    function increaseAllowance(address spender, uint256 addedAmount) public virtual returns (bool);
+}
+
+
+contract Router {
+    using SafeMath for uint256;
+
+    event Pledge(uint256 numCampaign, address addr, uint256 amount);
+    event Unpledge(uint256 numCampaign, address addr, uint256 amount);
+    event Rebase(uint256 numCampaign, treeSold, reserveTokenReceived);
+    event WithdrawToken(address token, address to, uint256 amount);
+    event SetReserveToken(address token);
+
+    address constant private TREE = 0xCE222993A7E4818E0D12BC56376c5a60f92A5783;
+    address constant private RESERVE = 0x390a8Fb3fCFF0bB0fCf1F91c7E36db9c53165d17;
+    address constant private DAI = 0x6B175474E89094C44Da98b954EedeAC495271d0F;
+
+    address public gov;
+    IERC20 public tree = IERC20(TREE);
+    IERC20 public reserveToken = IERC20(DAI);
+
+    struct Campaign {
+        uint256 totalPledged;
+        uint256 numPledgers;
+        uint256 treeSold;
+        mapping (uint256 => address) pledgers;
```

Comment:
Is this needed, or can we just check for zero balances on the next mapping? I couldn't see a real use for the pledgerId

Reply:
`pledgerId` is used when paying out TREE to pledgers, as there's no way to loop through a mapping.  Another option is making an array that holds objects like:
```Solidity
Struct pledger {
    address addr;
    uint256 amount;
}
```
and looping through the array using array.length.  Thoughts?

Reply:
I totally missed that use for `pledgerId`. Yes, looping through the array of struct is fine, and what you were doing before may have also been fine.  I will take another look.